text/0481-component-templates-co-location.md

```diff
@@ -0,0 +1,647 @@
+
+// GOOD: do this instead!
+export default class extends MyParentComponent {}
+```
+
+Most cases of this problem can be linted against easily.
+
+### Template-only components
+
+For template-only components, we propose to introduce the following low-level
+API:
+
+* The `templateOnlyComponent` function takes no arguments and produces a unique
+  value that can be used to represent a template-only component.
+
+Again, this function, though public, is not intended to be called by users
+directly. It is primarily used in the output emitted by build tools. When a
+template is found in `{app,addon}/components` but without a corresponding
+JavaScript file, the build output will be something similar to the following:
+
+```js
+// app/components/foo-bar.js
+
+import Component, { setComponentTemplate } from '@ember/component';
+import templateOnlyComponent from '@ember/component/template-only';
+
+// output of precompile("foo bar!")
+const TEMPLATE = {
+  id: "...",
+  block: "...",
+  meta: { moduleName: "app/components/foo-bar" }
+};
+
+const CLASS = templateOnlyComponent();
+
+export default setComponentTemplate(TEMPLATE, CLASS);
+```
+
+In addition to build tooling, addon authors may also find this function
+useful. Currently, the only way to create a "true" template-only component
+is by enabling an optional feature in the app. Since addons cannot assume the
+value of that flag in the consuming app, they currently cannot take advantage
+of the feature. By providing this function, addon authors can work around this
+problem by explicitly defining a JavaScript file with `templateOnlyComponent()`
+as the default export.
+
+On the other hand, app developers, should _not_ have any reason to use this
+function directly in their app, since they could just as easily enable the
+optional feature across the board.
+
+### Codemod
+
+A codemod will be provided to seamlessly migrate component templates.
+
+Such a "codemod" will essentially just merges `app/templates/components` into
+`app/components`. For a quick "taste test" of what the resulting tree will look
+like in your app, the following commands give a close approximation:
+
+```bash
+$ rsync --archive --remove-source-files app/templates/components/ app/components/
+$ rm -rf app/templates/components/
+```
+
+Of course, the resulting output won't "work", but it can be useful for getting
+a sense of what it's like to work with the new layout on editors, Github, etc.
+
+### Generator
+
+We propose to make some updates to the components generator to accompany this
+change.
+
+1. It should accept a `--component-class` option. This controls which base
+   class is used for that component and whether native classes are used. The
+   legal values for this option are `@glimmer/component` (aliased as `-gc`),
+   `@ember/component` (aliased as `-ec`), `@ember/component/template-only`
+   (aliased as `-tc`) or an empty string (aliased as `--no-component-class` and
+   `-nc`).
```

Comment:
Will we be referring to `@ember/component`s as "classic components" once Octane lands? If so, perhaps we should use `-cc` instead of `-ec` for `@ember/component`? This would also mirror the `-cs` classic structure below, "classic" would then consistently mean the old way of doing things.